amethyst_assets/src/source/http.rs

```diff
@@ -0,0 +1,92 @@
+use js_sys::Uint8Array;
+use std::path::{Path, PathBuf};
+use web_sys::{XmlHttpRequest, XmlHttpRequestResponseType};
+
+#[cfg(feature = "profiler")]
+use thread_profiler::profile_scope;
+
+use amethyst_error::{format_err, Error, ResultExt};
+
+use crate::{error, source::Source};
+
+/// HTTP source.
+///
+/// Loads assets inside web worker using XmlHttpRequest.
+/// Used as a default source for WASM target.
+#[derive(Debug)]
+pub struct HTTP {
+    loc: PathBuf,
+}
+
+impl HTTP {
+    /// Creates a new http source.
+    pub fn new<P>(loc: P) -> Self
+    where
+        P: Into<PathBuf>,
+    {
+        HTTP { loc: loc.into() }
+    }
+
+    fn path(&self, s_path: &str) -> PathBuf {
+        let mut path = self.loc.clone();
+        path.extend(Path::new(s_path).iter());
+
+        path
+    }
+}
+
+impl Source for HTTP {
+    fn modified(&self, _path: &str) -> Result<u64, Error> {
+        #[cfg(feature = "profiler")]
+        profile_scope!("http_modified_asset");
+
+        // Unimplemented. Maybe possible to tie into webpack hot module reloading?
+        Ok(0)
+    }
+
+    fn load(&self, path: &str) -> Result<Vec<u8>, Error> {
+        #[cfg(feature = "profiler")]
+        profile_scope!("http_load_asset");
+
+        let path = self.path(path);
+        let path_str = path
+            .to_str()
+            .ok_or_else(|| format_err!("Path contains non-unicode characters: {:?}", path))
+            .with_context(|_| error::Error::Source)?;
+
+        let xhr = XmlHttpRequest::new()
+            .map_err(|_| format_err!("Failed to construct XmlHttpRequest"))
+            .with_context(|_| error::Error::Source)?;
+
+        // Synchronous GET request. Should only be run in web worker.
+        xhr.open_with_async("GET", path_str, false).unwrap();
```

Comment:
can the `unwrap()`s be replaced with `?`s? Just a bit neater

Reply:
Everything in `web_sys` returns `JsValue` as error, so they have to be mapped. Some of the javascript functions (such as this one) do not return any error, but the `Result` return type is still used for the rust API. So I don't know a case where this could panic. Bugs in browser/wasm? Should I add error map even though it could be dead code?

Reply:
> Bugs in browser/wasm? Should I add error map even though it could be dead code?

haha, my conservative self always maps errors instead of unwraps -- wary of the 1/1000 case where it could save time debugging. Admittedly it isn't always pragmatic. Am happy to take the unreachable unwraps :v: